### PR TITLE
Tests to verify proper life-cycle of `RecyclerPool` implementations, from databind perspective

### PR DESCRIPTION
**Pull request by mariofusco**
Reproducer for the problem reported in https://github.com/FasterXML/jackson-databind/issues/4321

Also contains a small fix, closing the `SegmentedStringWriter` mistakenly left open in the `ObjectMapper`.

### CODE REVIEW

File: src/test/java/com/fasterxml/jackson/databind/util/BufferRecyclersDatabindTest.java [hunk 1]
```diff
@@ -92,14 +100,18 @@ public void testGeneratorWithDequeuPool() throws Exception {
 
     public void testGeneratorWithLockFreePool() throws Exception {
         _testGenerator(JsonRecyclerPools.newLockFreePool());
-        _testGenerator(JsonRecyclerPools.sharedLockFreePool());
+//        _testGenerator(JsonRecyclerPools.sharedLockFreePool());
```
Comment (cowtowncoder) [inline src/test/java/com/fasterxml/jackson/databind/util/BufferRecyclersDatabindTest.java:103]: Is this intentional?

Comment (mariofusco) [reply src/test/java/com/fasterxml/jackson/databind/util/BufferRecyclersDatabindTest.java:103]: No, sorry. I commented out temporarily to investigate that test and then forgot to uncomment.

Comment (cowtowncoder) [reply src/test/java/com/fasterxml/jackson/databind/util/BufferRecyclersDatabindTest.java:103]: Makes sense, just wanted to double-check.

**Review by cowtowncoder — APPROVED** [review]
Should be useful additional test, now passing, will merge.

**Pull request merged**